src/crimson/osd/main.cc

```diff
@@ -0,0 +1,103 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <iostream>
+
+#include <seastar/core/app-template.hh>
+#include <seastar/core/thread.hh>
+
+#include "common/ceph_argparse.h"
+#include "crimson/common/config_proxy.h"
+#include "crimson/net/SocketMessenger.h"
+
+#include "osd.h"
+
+using config_t = ceph::common::ConfigProxy;
+
+void usage(const char* prog) {
+  std::cout << "usage: " << prog << " -i <ID>" << std::endl;
+  generic_server_usage();
+}
+
+int main(int argc, char* argv[])
+{
+  std::vector<const char*> args{argv + 1, argv + argc};
+  if (ceph_argparse_need_usage(args)) {
+    usage(argv[0]);
+    return EXIT_SUCCESS;
+  }
+  std::string cluster;
+  std::string conf_file_list;
+  // ceph_argparse_early_args() could _exit(), while local_conf() won't ready
+  // until it's started. so do the boilerplate-settings parsing here.
+  auto init_params = ceph_argparse_early_args(args,
+                                              CEPH_ENTITY_TYPE_OSD,
+                                              &cluster,
+                                              &conf_file_list);
+  // TODO: add heartbeat
+  seastar::app_template app;
+  // talk with osd
+  ceph::net::SocketMessenger cluster_msgr("cluster", getpid());
+  // talk with mon/mgr
+  ceph::net::SocketMessenger client_msgr("client", getpid());
+  seastar::sharded<OSD> osd;
+
+  using ceph::common::sharded_conf;
+  using ceph::common::sharded_perf_coll;
+  using ceph::common::local_conf;
+
+  args.insert(begin(args), argv[0]);
+  try {
+    return app.run_deprecated(args.size(), const_cast<char**>(args.data()), [&] {
+      seastar::engine().at_exit([] {
+        return sharded_conf().stop();
+      });
+      seastar::engine().at_exit([] {
+        return sharded_perf_coll().stop();
+      });
+      seastar::engine().at_exit([&] {
+        return client_msgr.shutdown();
+      });
+      seastar::engine().at_exit([&] {
+       return osd.stop();
+      });
+      return sharded_conf().start(init_params.name, cluster).then([] {
+        return sharded_perf_coll().start();
+      }).then([&conf_file_list] {
+        return local_conf().parse_config_files(conf_file_list);
+      }).then([&] {
+        auto& conf = local_conf();
+        const auto whoami = std::stoi(conf->name.get_id());
+        for (ceph::net::SocketMessenger& msgr : {std::ref(cluster_msgr),
+                                                 std::ref(client_msgr)}) {
+          msgr.set_myname(entity_name_t::OSD(whoami));
+          if (conf->ms_crc_data) {
+            msgr.set_crc_data();
+          }
+          if (conf->ms_crc_header) {
+            msgr.set_crc_header();
+          }
+        }
+        return osd.start_single(std::stoi(conf->name.get_id()),
+                                std::ref(cluster_msgr),
```

Comment:
`OSD` currently just *uses* these messengers. I guess it could *own* them and control their lifetimes as well (no messenger sharing; `std::unique_ptr<Messenger>` in OSD's ctor). I have [a patch](https://github.com/rzarzynski/ceph/commit/0dacca970b8e3507410ec4b388fdfffae5f47566) doing that for `MonClient` in the sharded world. Maybe similar (but much simpler) thing could be made here.

Anyway, this is definitely not a blocker.

Reply:
done.

Reply:
> I would move the rest of those things, if possible, to use one-step construction approach as well 

let's leave this for anther PR.